tests/utilities/test_auto_restart.py

```diff
@@ -970,3 +972,172 @@ def test_dataset_rng_states_restart_with_lightning(tmpdir, dataset_classes, mult
     for w0, w1 in zip(weights0, weights1):
         assert w0 is not w1
         assert torch.allclose(w0, w1)
+
+
+class ValidationLoopTestModel(LightningModule):
+    def __init__(self, fail_on_dataloader: int = -1, val_check_interval: float = 0.5):
+        super().__init__()
+        self.layer = torch.nn.Linear(1, 2)
+        self.training_seen_batches = []
+        self.validation_seen_batches = defaultdict(list)
+        self.fail_on_dataloader = fail_on_dataloader
+        self.failing_batch_idx: Optional[int] = None
+        self.failing_dataloader_int: Optional[int] = None
+        self.val_check_interval = val_check_interval
+
+    def training_step(self, batch, batch_idx):
+        batch = batch["data"] if isinstance(batch, dict) else batch
+        self.training_seen_batches.append(torch.stack(batch) if isinstance(batch, list) else batch)
+        loss = sum(self.layer(b).sum() for b in batch)
+        return loss
+
+    def validation_step(self, batch, batch_idx, dataloader_int: int = 0):
+        loss = sum(self.layer(b).sum() for b in batch)
+
+        if self.trainer.sanity_checking:
+            return loss
+
+        # failure on first batch of dataloader_idx once global_step is 1
+        if (
+            self.fail_on_dataloader >= 0
+            and self.fail_on_dataloader == dataloader_int
+            and batch_idx == 1
+            and (self.trainer.global_step == 3 if self.val_check_interval == 1.0 else self.trainer.global_step == 1)
+        ):
+            self.failing_batch_idx = batch_idx
+            self.failing_dataloader_int = dataloader_int
+            raise CustomException
+
+        self.validation_seen_batches[dataloader_int].append(batch)
+        return loss
+
+    def configure_optimizers(self):
+        return torch.optim.SGD(self.layer.parameters(), lr=0.1)
+
+
+@mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "1"})
+@RunIf(min_torch="1.7.0")
+@pytest.mark.parametrize(
+    "dataset_classes",
+    [
+        [[RandomGetItemDataset], [RandomGetItemDataset]],
+        [[RandomGetItemDataset], [RandomGetItemDataset, RandomGetItemDataset]],
+    ],
+)
+@pytest.mark.parametrize("val_check_interval", [0.5, 1.0])
+def test_auto_restart_within_validation_loop(dataset_classes, val_check_interval, tmpdir):
```

Comment:
IMO this should be split into two tests:

one in `tests/loops/...` to check the progress counters from checkpoints

the other here to check the dataloader state etc

Reply:
Done !

Reply:
Done !